guides/rails_guides/markdown/renderer.rb

```diff
@@ -66,16 +66,27 @@ def convert_footnotes(text)
           end
         end
 
+        def sanitize_lexer(code_type)
+          language = lexer_language(code_type)
+          lexer = ::Rouge::Lexer.find_fancy(language)
+          # Revert to plain text when Rouge doesn't recognize the language
+          if lexer.nil?
+            language = "plaintext"
+            lexer = ::Rouge::Lexer.find(language)
+          end
+          [language, lexer]
+        end
+
         def lexer_language(code_type)
           case code_type
-          when "css", "js", "html", "ruby", "sql", "yaml"
-            code_type
-          when "erb", "html+erb"
+          when "html+erb"
             "erb"
           when "bash"
             "console"
-          else
+          when nil
             "plaintext"
+          else
+            code_type
```

Comment:
An alternative to the `sanitize_lexer` method might be to instead always return a valid language from `lexer_language`:

```suggestion
            ::Rouge::Lexer.find(code_type) ? code_type : "plaintext"
```

Reply:
@jonathanhefner Yeah this was my first thought, but `find` doesn't support passing options such as the prompt or `guess`. I could strip these before calling find. I just thought as we need to call fancy_find anyway, the way I'm doing it we only need to call it once so long as the type is valid.

I'm happy to switch to doing that though if you would still recommend it?

Reply:
I'm not sure I follow what you mean...?  What I meant was `Rouge::Lexer.find(code_type)` can be used as `Rouge::Lexer.registry.key?(code_type)` would be (since `Rouge::Lexer.registry` is private).  So `::Rouge::Lexer.find(code_type) ? code_type : "plaintext"` as the final case of `lexer_language` would always return a valid language.  Thus the `block_code` method could remain the same, and `sanitize_lexer` wouldn't be necessary.

Reply:
I've made some changes based on what you suggested. It's probably best to look at the overall changes but I didn't squash the commit yet so you could see the changes since the last commit.

Rouge allows you to pass parameters to some of the lexers such as the Console lexer to change what it identifies as the prompt and what lexer the rest of the line is then passed onto. This allows you to have lines starting with an `irb`, `mysql>` etc but not include that prompt in the syntax highlighting.

So a valid code_type (just one possible example) is `console?prompt=irb>&lang=ruby`. Passing this to `::Rouge::Lexer.find` would result in a `nil` even though it's valid.

An example of an area that could have syntax highlighting added thanks to this: https://edgeguides.rubyonrails.org/debugging_rails_applications.html#verbose-query-logs

Reply:
I would be interested to see other places in the guides that would benefit from this.  If we do want to leverage this feature, I think it would be good to do the `split` prior to the `case` statement, yes?

Reply:
Yeah, the case statement is for handling legacy syntax languages still used in the document. We should probably move away from those and towards the ones supported by Rouge. I wanted to keep this PR focused on adding back the missing language support. I have other fixes I wanted to make in other PRs, but I wanted to keep this one simple to review, especially as it's my first to Rails.

> I would be interested to see other places in the guides that would benefit from this.

I think this was why the original author used `fancy_find` over `find`. It's difficult to link to the exact documentation sections as there is sometimes no anchor, but here are some examples:
https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#reloading
https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants_classic_mode.html#autoloading-within-singleton-classes (scroll up a tiny bit)
https://edgeguides.rubyonrails.org/development_dependencies_install.html#database-configuration (output is treated as if it's Ruby and the same is true for many of the examples after that point)
https://edgeguides.rubyonrails.org/command_line.html#bin-rails-server (The prompt is treated as if it's ruby)
There's a really good example about mid-way through https://edgeguides.rubyonrails.org/command_line.htm. Search for "bin/rails generate model" and you will see that the entire block is treated as if it's a shell script and there's a similar example two down from that where the table creation time is highlighted for the same reason.

I going to stop there but there are plenty of other examples. It is also an issue for people creating new code blocks and running into similar annoyances.

Thanks again for your time. :)

Reply:
Thank you for all those examples!  It looks like most are `bash` related.  What if instead of adding the `?` params to the code fences in the guides, we add them to the `bash` case in `lexer_language`, i.e. `"bash"` => `"console?..."`?

For the IRB-related examples, it looks like there is an [IRB lexer](https://github.com/rouge-ruby/rouge/blob/v3.20.0/lib/rouge/lexers/irb.rb), so it's just a matter of changing the code fences and using a standard prompt in the IRB examples.  (If we don't like Rouge's choice of prompt, we could add an `"irb"` => `"irb?..."` case.)  Like you said though, I think that would be good for a follow-up PR.

And the [SQL console example](https://edgeguides.rubyonrails.org/development_dependencies_install.html#database-configuration) looks like it is mostly fixed by changing the code fence from `bash` to `sql`, as you did.  If we wanted to fix it further, we could possibly extract the shell command from the code snippet into the paragraph above it.  Or if this is a recurring problem, we could consider adding a `"dbconsole"` => `"console?lang=sql..."` case in `lexer_language`.

> We should probably move away from those and towards the ones supported by Rouge.

The risk in doing this is that we may need to replace Rouge some day.  Also, some code fences are commonly supported elsewhere (e.g. `bash` on GitHub), and so I feel it would be good to still support them in the guides.

Reply:
> The risk in doing this is that we may need to replace Rouge some day.

Well anything could happen, but I was thinking about adjusting the code to use what we currently have.

> Also, some code fences are commonly supported elsewhere (e.g. bash on GitHub), and so I feel it would be good to still support them in the guides.

I feel a lot of things are currently labelled as bash when the are not Bash. But you are probably right about the syntax highlighting in areas that are not Rouge, so fair enough.

> Thank you for all those examples! It looks like most are bash related. What if instead of adding the ? params to the code fences in the guides, we add them to the bash case in lexer_language, i.e. "bash" => "console?..."?

Yeah there's also the `debug` param which allows you to debug the lexer, but I guess this could also be temporarily added in the `lexer_language` method by any dev who wants to debug.  I can understand that you probably want to lock this list down. It's just that there is currently lots of inconsistencies in things such as prompts used the final syntax needed. But I agree that we could build on the lexer_language case for these instances.

For now, to just add back support for the languages we lost and add all the other languages supported, I'm going to use your original suggestion of just using `find`, which will not allow prompt options.

Thanks for your persistence. 😄

Reply:
This should just have docs label and none of the others. I accidentally rebase on the wrong remote (fixed now). Any chance you could remove the other labels?